pkg/mappings/mappings.go

```diff
@@ -0,0 +1,173 @@
+// Copyright Contributors to the Open Cluster Management project
+
+package mappings
+
+import (
+	"bytes"
+	_ "embed"
+	"encoding/json"
+	"io"
+	"net/http"
+
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/kubectl/pkg/scheme"
+	"k8s.io/kubernetes/pkg/apis/apidiscovery"
+)
+
+// APIMapping stores information required for mapping between GroupVersionKinds
+// and GroupVersionResources, as well as whether the API is namespaced.
+type APIMapping struct {
+	Group    string
+	Version  string
+	Kind     string
+	Singular string
+	Plural   string
+	Scope    string
+}
+
+// MapperFrom takes a list of APIMappings and uses them to populate a
+// DefaultRESTMapper which can be used locally without connecting to a Discovery
+// client.
+func MapperFrom(mappings []APIMapping) meta.RESTMapper {
+	mapper := meta.NewDefaultRESTMapper(scheme.Scheme.PreferredVersionAllGroups())
+
+	for _, mapping := range mappings {
+		gv := schema.GroupVersion{
+			Group:   mapping.Group,
+			Version: mapping.Version,
+		}
+
+		if mapping.Scope == "Cluster" {
+			mapper.AddSpecific(
+				gv.WithKind(mapping.Kind),
+				gv.WithResource(mapping.Plural),
+				gv.WithResource(mapping.Singular),
+				clusterScoped,
+			)
+		} else if mapping.Scope == "Namespaced" {
+			mapper.AddSpecific(
+				gv.WithKind(mapping.Kind),
+				gv.WithResource(mapping.Plural),
+				gv.WithResource(mapping.Singular),
+				namespaced,
+			)
+		}
+	}
+
+	return mapper
+}
+
+//go:embed default-mappings.json
+var defaultMappingsJSON []byte
+
+// DefaultMapper returns a RESTMapper initialized on default mappings that were
+// embedded into this package. It should include built-in Kubernetes types.
+func DefaultMapper() (meta.RESTMapper, error) {
+	mappings := []APIMapping{}
+
+	if err := json.Unmarshal(defaultMappingsJSON, &mappings); err != nil {
+		return nil, err
+	}
+
+	return MapperFrom(mappings), nil
+}
+
+// GenerateMappings connects to a Kubernetes cluster and discovers the available
+// api-resources, printing out that information as a JSON list of APIMappings.
+// The cluster connected to follows the usual conventions, eg it can be set with
+// the KUBECONFIG environment variable. This function is meant to be used inside
+// of a cobra-style CLI.
+func GenerateMappings(cmd *cobra.Command, _ []string) error {
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	clientConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &clientcmd.ConfigOverrides{})
+
+	kubeConfig, err := clientConfig.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	cli, err := rest.HTTPClientFor(kubeConfig)
+	if err != nil {
+		return err
+	}
+
+	apiMappings := []APIMapping{}
+
+	for _, endpoint := range []string{"/api", "/apis"} {
```

Comment:
Is there a reason to do this instead of leveraging `GroupsAndMaybeResources` on the discovery client?

Reply:
Only because I wasn't familiar with that method, and I was somewhat clumsily recreating what I saw `kubectl` do :laughing: